contrib/codegen-tools/codegen/src/main/ops/org/nd4j/codegen/mixins/Mixins.kt

```diff
@@ -94,31 +94,67 @@ val scalar = Mixin("scalar"){
 val reduce = Mixin("reduce"){
     legacy = true
     Input(DataType.NUMERIC, "in") { description = "Input variable" }
+    Arg(DataType.BOOL,"keepDims"){"Whether to keep the original  dimensions or produce a shrunk array with less dimensions"}
     Arg(DataType.INT, "dimensions"){ count = AtLeast(0); description = "Dimensions to reduce over. If dimensions are not specified, full array reduction is performed" }
     Output(DataType.NUMERIC, "output"){ description = "Reduced array of rank (input rank - num dimensions)" }
 }
 
+val reduceVariableDimensions = Mixin("reduceVariable") {
+    legacy = true
+    Input(DataType.NUMERIC, "in") { description = "Input variable" }
+    Input(DataType.NUMERIC, name = "dimensions"){description = "Dimensions to reduce along"; defaultValue = null }
+    Arg(DataType.BOOL,"keepDims"){"Whether to keep the original  dimensions or produce a shrunk array with less dimensions"}
+    Output(DataType.NUMERIC, "output"){ description = "Reduced array of rank (input rank - num dimensions)" }
+}
+
 val reduceFloating = Mixin("reduceFloating"){
     useMixin(reduce)
-    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.floating"
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.floating.custom"
 }
 
+val reduceFloatingVariable = Mixin("reduceFloatingVariable"){
+    useMixin(reduceVariableDimensions)
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.floating.custom"
+}
+
+
 val reduceSame = Mixin("reduceSame"){
     useMixin(reduce)
-    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.same"
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.same.custom"
+}
+
+val reduceSameVariable = Mixin("reduceSameVariable"){
+    useMixin(reduceVariableDimensions)
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.same.custom"
 }
 
 val reduceLong = Mixin("reduceLong"){
     useMixin(reduce)
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.longer.custom"
+}
+
+val reduceLongVariable = Mixin("reduceLongVariable"){
+    useMixin(reduceVariableDimensions)
     javaPackage = "org.nd4j.linalg.api.ops.impl.reduce.longer"
 }
 
 val reduce3 = Mixin("reduce3"){
     legacy = true
-    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce3"
+    javaPackage = "org.nd4j.linalg.api.ops.impl.reduce3.custom"
+    Input(DataType.NUMERIC, "x") { description = "Input variable x" }
+    Input(DataType.NUMERIC, "y") { description = "Input variable y" }
+    Arg(DataType.BOOL,"keepDims",{description = "Whether to preserve original dimensions or not"})
+    Arg(DataType.BOOL,"isComplex",{description = "Depending on the implementation, such as distance calculations, this can determine whether all distance calculations for all points should be done."; defaultValue = false})
+    val dims = Arg(DataType.INT, "dimensions"){ count = AtLeast(1); isVargarg = true; description = "Dimensions to reduce over. If dimensions are not specified, full array reduction is performed" }
```

Comment:
Does it make sense to require at least one? I guess that would be to enforce that people are aware that it is going to go over at least one dimension? 

In any case that looks like it may break some existing code, as you are adding additional required arguments here.